File: backend/src/main/kotlin/net/animeta/backend/service/ChartService.kt
package net.animeta.backend.service

import com.google.common.cache.Cache
import com.google.common.cache.CacheBuilder
import com.mysema.commons.lang.CloseableIterator
import com.querydsl.core.types.Expression
import com.querydsl.core.types.Projections
import com.querydsl.jpa.HQLTemplates
import com.querydsl.jpa.impl.JPAQuery
import net.animeta.backend.chart.ChartItem
import net.animeta.backend.chart.ChartRange
import net.animeta.backend.chart.diff
import net.animeta.backend.chart.ranked
import net.animeta.backend.model.History
import net.animeta.backend.model.QHistory.history
import net.animeta.backend.model.QRecord
import net.animeta.backend.repository.UserRepository
import net.animeta.backend.repository.WorkRepository
import net.animeta.backend.serializer.WorkSerializer
import org.springframework.stereotype.Service
import java.sql.Timestamp
import java.time.ZoneId
import java.util.*
import java.util.concurrent.TimeUnit
import javax.persistence.EntityManager

@Service
class ChartService(val entityManager: EntityManager,
                   val workRepository: WorkRepository,
                   val userRepository: UserRepository,
                   val workSerializer: WorkSerializer) {
    data class ChartItemWork(val id: Int, val title: String, val image_url: String?)
    data class ChartItemUser(val username: String)

    private val maxLimit = 100
    private val popularWorksCache: Cache<Optional<ChartRange>, List<ChartItem<ChartItemWork>>> =
            CacheBuilder.newBuilder()
                    .maximumSize(3) // {weekly, monthly, overall}
                    .expireAfterWrite(1, TimeUnit.HOURS)
                    .build()
    private val activeUsersCache: Cache<Optional<ChartRange>, List<ChartItem<ChartItemUser>>> =
            CacheBuilder.newBuilder()
                    .maximumSize(3) // {weekly, monthly, overall}
                    .expireAfterWrite(1, TimeUnit.HOURS)
                    .build()

    fun defaultTimeZone() = ZoneId.of("Asia/Seoul")

    fun getPopularWorks(range: ChartRange?, limit: Int): List<ChartItem<ChartItemWork>> {
        return popularWorksCache.get(Optional.ofNullable(range)) { getPopularWorksUncached(range) }.take(limit)
    }

    fun getPopularWorksUncached(range: ChartRange?): List<ChartItem<ChartItemWork>> {
        val chart = if (range != null) {
            iteratePopularWorks(range).use { a ->
                iteratePopularWorks(range.prev()).use { b ->
                    diff(ranked(a.asSequence()).take(maxLimit), ranked(b.asSequence()))
                }
            }
        } else {
            iteratePopularWorks(range).use {
                ranked(it.asSequence()).take(maxLimit).toList()
            }
        }
        val works = workRepository.findAll(chart.map { it.`object` }).associateBy { it.id }
        return chart.map { it.map { id ->
            val work = works[id]!!
            ChartItemWork(work.id!!, work.title, workSerializer.getImageUrl(work))
        } }
    }

    private fun iteratePopularWorks(range: ChartRange?): CloseableIterator<Pair<Int, Int>> {
        if (range != null) {
            val factor = history.user.countDistinct()
            val instantRange = range.instantRange(defaultTimeZone())
            return JPAQuery<History>(entityManager, HQLTemplates.DEFAULT)
                    .select(Projections.constructor(Pair::class.java, history.work.id, factor) as Expression<Pair<Int, Int>>)
                    .from(history)
                    .where(history.updatedAt.between(Timestamp.from(instantRange.lowerEndpoint()), Timestamp.from(instantRange.upperEndpoint())))
                    .groupBy(history.work.id)
                    .orderBy(factor.desc())
                    .having(factor.gt(1))
                    .iterate()
        } else {
            val factor = QRecord.record.count()
            return JPAQuery<History>(entityManager, HQLTemplates.DEFAULT)
                    .select(Projections.constructor(Pair::class.java, QRecord.record.work.id, factor) as Expression<Pair<Int, Int>>)
                    .from(QRecord.record)
                    .groupBy(QRecord.record.work.id)
                    .orderBy(factor.desc())
                    .having(factor.gt(1))
                    .iterate()
        }
    }

    fun getActiveUsers(range: ChartRange?, limit: Int): List<ChartItem<ChartItemUser>> {
        return activeUsersCache.get(Optional.ofNullable(range)) { getActiveUsersUncached(range) }.take(limit)
    }

    fun getActiveUsersUncached(range: ChartRange?): List<ChartItem<ChartItemUser>> {
        val chart = if (range != null) {
            iterateActiveUsers(range).use { a ->
                iterateActiveUsers(range.prev()).use { b ->
                    diff(ranked(a.asSequence()).take(maxLimit), ranked(b.asSequence()))
                }
            }
        } else {
            iterateActiveUsers(range).use {
                ranked(it.asSequence()).take(maxLimit).toList()
            }
        }
        val users = userRepository.findAll(chart.map { it.`object` }).associateBy { it.id }
        return chart.map { it.map { id ->
            val user = users[id]!!
            ChartItemUser(user.username)
        } }
    }

    private fun iterateActiveUsers(range: ChartRange?): CloseableIterator<Pair<Int, Int>> {
        val factor = history.user.count()
        val query = JPAQuery<History>(entityManager, HQLTemplates.DEFAULT)
                .select(Projections.constructor(Pair::class.java, history.user.id, factor) as Expression<Pair<Int, Int>>)
                .from(history)
        if (range != null) {
            val instantRange = range.instantRange(defaultTimeZone())
            query.where(history.updatedAt.between(Timestamp.from(instantRange.lowerEndpoint()), Timestamp.from(instantRange.upperEndpoint())))
        }
        return query.groupBy(history.user.id)
                .orderBy(factor.desc())
                .iterate()
    }
}

File: backend-ts/src/services/chart.service.ts
import { CACHE_MANAGER, Inject, Injectable } from "@nestjs/common";
import { ChartItem, ChartItemWork } from "shared/types";
import { SundayStartWeekRange } from "src/utils/sunday_start_week_range";
import { Temporal } from "@js-temporal/polyfill";
import { Connection } from "typeorm";
import { Work } from "src/entities/work.entity";
import { History } from "src/entities/history.entity";
import { diff, ranked } from "src/utils/chart";
import { WorkService } from "./work.service";
import { Cache } from "cache-manager";

interface ChartRange {
  prev(): ChartRange
  startDate(): Temporal.PlainDate
  endDate(): Temporal.PlainDate
}

@Injectable()
export class ChartService {
  private readonly maxLimit = 100
  private readonly defaultTimeZone = Temporal.TimeZone.from('Asia/Seoul')

  constructor(
    private connection: Connection,
    private workService: WorkService,
    @Inject(CACHE_MANAGER) private cache: Cache,
  ) {}

  async getWeeklyWorks(limit: number): Promise<Array<ChartItem<ChartItemWork>>> {
    const range = SundayStartWeekRange.now(this.defaultTimeZone)
    const cacheKey = `getWeeklyWorks:${range.sunday}`
    const result = await this.cache.wrap<Array<ChartItem<ChartItemWork>>>(
      cacheKey,
      () => this.getPopularWorksUncached(range),
      { ttl: 60 * 60 }
    )
    return result.slice(0, limit)
  }

  private async getPopularWorksUncached(range: ChartRange): Promise<Array<ChartItem<ChartItemWork>>> {
    const [currentChart, prevChart] = await Promise.all([
      this.getPopularWorks(range),
      this.getPopularWorks(range.prev())
    ])
    const chart = diff(ranked(currentChart).slice(0, this.maxLimit), ranked(prevChart))
    const works = await Promise.all(chart.map(it => this.workService.get(it.object)))
    const worksById = new Map<number, Work>(works.map(it => [it.id, it]))
    return chart.map(item => {
      const work = worksById.get(item.object)!
      const newItem: ChartItem<ChartItemWork> = {
        ...item,
        object: {
          id: work.id,
          title: work.title,
          image_url: this.workService.getImageUrl(work),
          image_center_y: null,
        }
      }
      return newItem
    })
  }

  private async getPopularWorks(range: ChartRange): Promise<Array<[number, number]>> {
    const startInstant = range.startDate().toZonedDateTime({ timeZone: this.defaultTimeZone }).toInstant()
    const endInstant = range.endDate().add({ days: 1 }).toZonedDateTime({ timeZone: this.defaultTimeZone }).toInstant()
    const result = await this.connection.createQueryBuilder()
      .from(History, 'h')
      .select('h.work_id', 'workId')
      .addSelect('COUNT(DISTINCT h.user_id)', 'factor')
      .where('h.updated_at BETWEEN :start AND :end', { start: new Date(startInstant.epochMilliseconds), end: new Date(endInstant.epochMilliseconds) })
      .groupBy('h.work_id')
      .having('COUNT(DISTINCT h.user_id) > 1')
      .orderBy('factor', 'DESC')
      .limit(this.maxLimit)
      .getRawMany()
    return result.map(it => [it.workId, Number(it.factor)])
  }
}
